pkg/planner/core/preprocess.go

```diff
@@ -1839,14 +1839,39 @@ func (p *preprocessor) hasAutoConvertWarning(colDef *ast.ColumnDef) bool {
 	return false
 }
 
-func tryLockMDLAndUpdateSchemaIfNecessary(ctx context.Context, sctx base.PlanContext, dbName pmodel.CIStr, tbl table.Table, is infoschema.InfoSchema) (table.Table, error) {
+func tryLockMDLAndUpdateSchemaIfNecessary(ctx context.Context, sctx base.PlanContext, dbName pmodel.CIStr, tbl table.Table, is infoschema.InfoSchema) (retTbl table.Table, err error) {
+	skipLock := false
+	shouldLockMDL := false
+	var lockedID int64
+
+	defer func() {
```

Comment:
Why it's defined as a defer instead of directly checking it?

Reply:
Because we need to check `retTbl`. The only way to check the `retTbl` without modifying every `return` statement is to check it in `defer` :thinking:  .

The `retTbl` can be different with the `tbl` in argument, so we cannot check it only once at the entry of this function.